pkg/cmd/infra/router/template_test.go

```diff
@@ -0,0 +1,132 @@
+			expectError:        true,
+			expectErrorMessage: "invalid HTTP header input specification",
+		},
+		{
+			description: "should percent decode the multiple response header values",
+			inputValue:  "X-Frame-Options:DENY:Set,X-XSS-Protection:1%3Bmode%3Dblock:Set,x-forwarded-client-cert:%25%7B%2BQ%7D%5Bssl_c_der%2Cbase64%5D:Set",
+			expectedValue: []templateplugin.HTTPHeader{
+				{Name: "X-Frame-Options", Value: "'DENY'", Action: "Set"},
+				{Name: "X-XSS-Protection", Value: "'1;mode=block'", Action: "Set"},
+				{Name: "x-forwarded-client-cert", Value: "'%{+Q}[ssl_c_der,base64]'", Action: "Set"},
+			},
+			expectError: false,
+		},
+		{
+			description: "should percent decode the multiple request header values",
+			inputValue:  "Accept:text%2Fplain%2C+text%2Fhtml:Set,Accept-Encoding:Delete",
+			expectedValue: []templateplugin.HTTPHeader{
+				{Name: "Accept", Value: "'text/plain, text/html'", Action: "Set"},
+				{Name: "Accept-Encoding", Action: "Delete"},
+			},
+			expectError: false,
+		},
+		{
+			description: "should percent decode the multiple header values with simple non encoded strings",
+			inputValue:  "X-Frame-Options:DENY:Set",
+			expectedValue: []templateplugin.HTTPHeader{
+				{Name: "X-Frame-Options", Value: "'DENY'", Action: "Set"},
+			},
+			expectError: false,
+		},
+		{
+			description:        "When Action is `Set` it should not percent decode the incorrectly encoded value part Eg: percent is not followed by two hexadecimal digits",
+			inputValue:         "x-forwarded-client-cert:%25%7B%2BQ%7D%5Bssl_c_der%2Cbase64%5:Set",
+			expectError:        true,
+			expectedValue:      nil,
+			expectErrorMessage: "failed to decode percent encoding",
+		},
+		{
+			description:        "Invalid Action when header input spec has 2 parts",
+			inputValue:         "Accept-Encoding:Delete%2",
+			expectError:        true,
+			expectedValue:      nil,
+			expectErrorMessage: "invalid action Delete%2",
+		},
+		{
+			description:        "Invalid Action when header input spec has 3 parts",
+			inputValue:         "X-Frame-Options:DENY:Bad",
+			expectedValue:      nil,
+			expectError:        true,
+			expectErrorMessage: "invalid action Bad",
+		},
+		{
+			description:        "Should not allow specifying headers with no action or value",
+			inputValue:         "X-Frame-Options",
+			expectError:        true,
+			expectErrorMessage: "invalid HTTP header input specification: ",
+		},
+		{
+			description:        "Should not allow blank values for headers",
+			inputValue:         "",
+			expectedValue:      nil,
+			expectError:        true,
+			expectErrorMessage: "encoded header string not present",
+		},
+		{
+			description:        "Should fail with incorrect header name when HTTP header input specification has 3 parts",
+			inputValue:         "X-Frame-Options**!=:DENY:Set",
+			expectedValue:      nil,
+			expectError:        true,
+			expectErrorMessage: "invalid HTTP header name: X-Frame-Options**!",
```

Comment:
Why doesn't this header name have the `=` at the end like the inputValue?

Reply:
I recommended some changes in your loop over the testCases.  https://github.com/openshift/router/pull/438/files#diff-a391430739b2cebfb45c7a0eb8c8b80910ba610cf34d7e9d80b08d005a368124R120 in particular, should not be the first case evaluated or you won't be able to compare the expected error message.

Reply:
I don't think your error messages are being checked.

Reply:
ack made the changes